akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/internal/ClusterActorSetImpl.scala

```diff
@@ -0,0 +1,178 @@
+import akka.cluster.sharding.typed.scaladsl.Entity
+import akka.cluster.sharding.typed.scaladsl.EntityContext
+import akka.cluster.sharding.typed.scaladsl.EntityTypeKey
+
+import scala.concurrent.duration.Duration
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] object ClusterActorSetImpl {
+  object EntityParent {
+    sealed trait Command
+    case object Ping extends Command with DeadLetterSuppression
+    case object Stop extends Command
+
+    def apply[T](
+        entityContext: EntityContext[Command],
+        factory: EntityId => Behavior[T],
+        childStopMessage: T): Behavior[Command] =
+      Behaviors.setup { context =>
+        var child: Option[ActorRef[T]] = None
+
+        def startChild(): Unit = {
+          val startedChild = context.spawn(factory(entityContext.entityId), "entity")
+          context.watch(startedChild)
+          child = Some(startedChild)
+        }
+
+        context.log.debug(s"Starting ClusterActorSet actor ${entityContext.entityId}")
+        startChild()
+
+        Behaviors
+          .receiveMessage[Command] {
+            case Stop =>
+              child match {
+                case Some(ref) =>
+                  ref ! childStopMessage
+                  waitingForChildToStop()
+                case None =>
+                  Behaviors.stopped
+              }
+
+            case Ping =>
+              Behaviors.same
+          }
+          .receiveSignal {
+            case (_, Terminated(_)) =>
+              // if the child stopped manually, but we are not being re-located to another node,
+              // we can fast-restart it instead of waiting for a ping
+              context.log.debug(s"ClusterActorSet actor ${entityContext.entityId} stopped, starting again")
+              startChild()
+              Behaviors.same
+          }
+      }
+
+    private def waitingForChildToStop(): Behavior[Command] = Behaviors.setup { context =>
+      // FIXME timeout and force termination if child doesn't stop?
+      Behaviors.receiveSignal {
+        case (_, Terminated(_)) =>
+          // if the child stopped manually, but we are not being re-located to another node,
+          // we can fast-restart it instead of waiting for a ping
+          context.log.debug(s"ClusterActorSet actor stopped")
+          Behaviors.stopped
+      }
+    }
+  }
+
+  object KeepAlivePinger {
+    sealed trait Event
+    case object Tick extends Event
+
+    def apply(
+        settings: ClusterActorSetSettings,
+        identities: Set[EntityId],
+        shardingRef: ActorRef[ShardingEnvelope[EntityParent.Command]]): Behavior[Event] =
+      Behaviors.setup { context =>
+        Behaviors.withTimers { timers =>
+          context.log.debug(
+            s"Starting ClusterActorSet KeepAlivePinger with ping interval ${settings.keepAliveInterval}")
+          // FIXME should we delay initial tick with random fraction of keepAliveInterval to avoid thundering herd?
```

Comment:
I don't think that is needed given the expected use case is a small number of them

Reply:
I also think that we need another message, eg: `StartId(id: String)`. When we receive the `Tick` we send `StartId(id)` to self of each id in `identities`. And we throttle.

Reply:
On the other hand I'm sure some want things to start as quickly as possible.
Maybe startup throttling should be done on the receiving side instead, specific to the read-side processor.

Reply:
Right, this should be an implementatino detail of the worker.